src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -2309,7 +2315,37 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
         ASR::symbol_t *stemp = s;
         s = ASRUtils::symbol_get_past_external(s);
 
-        if(ASR::is_a<ASR::Function_t>(*s)) {
+        if (ASR::is_a<ASR::GenericProcedure_t>(*s)) {
+            s_generic = stemp;
+            ASR::GenericProcedure_t *p = ASR::down_cast<ASR::GenericProcedure_t>(s);
+            int idx = ASRUtils::select_generic_procedure(args, *p, x.base.base.loc,
+                [&](const std::string &msg, const Location &loc) { throw SemanticError(msg, loc); });
+            s = p->m_procs[idx];
+            std::string remote_sym = ASRUtils::symbol_name(s);
+            std::string local_sym = ASRUtils::symbol_name(s);
+            if (ASR::is_a<ASR::ExternalSymbol_t>(*stemp)) {
+                local_sym = std::string(p->m_name) + "@" + local_sym;
```

Comment:
It's name mangling, so that the symbol can be looked up and reused with multiple calls (so it can't just be a unique symbol that is automatically generated by incrementing a unique counter), and at the same time does not shadow a possibly use defined function.

Reply:
I see, thanks!

Reply:
The only concern I have is that it could become verbose for each `FunctionCall` node in the ASR.

This will be generated for `bool(complex(0, 0))`:

ASR:
```
                  (FunctionCall 1 bool@__lpython_overloaded_7__bool 6 bool [
                     (FunctionCall 6 complex () [(ConstantInteger 0
                        (Integer 4 [])) (ConstantInteger 0
                        (Integer 4 []))] []
                        (Complex 8 [])
                        (ConstantComplex 0.000000 0.000000
                           (Complex 8 [])) ())] []
                     (Logical 1 [])
                     (ConstantLogical .false.
                        (Logical 1 [])) ()) ())
```
I haven't overloaded `complex` yet.

Reply:
Yes, instead of `__lpython_overloaded_7__bool` it would be `bool@__lpython_overloaded_7__bool`, that's the only difference. For `bool(complex(0, 0))` we should actually just create "ImplicitCast" (which should just be renamed to Cast) nodes in ASR directly, not call a function.